### PR TITLE
Add suggest-change-store.ts and db migration

### PR DESCRIPTION
**Pull request by andreas-unleash**
## About the changes

Creates the suggest-change-set table and starting store functionality `create` and `get`

Closes # .

### CODE REVIEW

File: src/lib/db/suggest-change-store.ts
Comment (thomasheartman) [inline src/lib/db/suggest-change-store.ts:49]: I had to go look this type up, so I might not have it completely correct. Does this make the `id` property optional on the incoming changeset? In that case, what happens to the `id` further down? If it's not there, does the DB set one automatically?

File: src/lib/types/model.ts
Comment (thomasheartman) [inline src/lib/types/model.ts:376]: I thought (at least) some of these options were required: definitely `project`, but probably also `changes` and the `created*` options. Why are they made optional here?

File: src/lib/types/stores.ts
Comment (thomasheartman) [inline src/lib/types/stores.ts:64]: This looks like a little typo 🦅 

```suggestion
    suggestChangeStore: ISuggestChangeStore;
```

**Review by thomasheartman — COMMENTED** [review]
Nice work! Looks like a good start to me, but I've got a couple of questions regarding typing. Would you mind having a look? Also, there doesn't seem to be any tests included at the moment: is there nothing to test?

File: src/lib/db/suggest-change-store.ts
Comment (andreas-unleash) [reply src/lib/db/suggest-change-store.ts:49]: yes the db should add it

File: src/lib/types/model.ts
Comment (andreas-unleash) [reply src/lib/types/model.ts]: resolved, leftover from developing. thanks!

File: src/lib/db/suggest-change-store.ts
Comment (thomasheartman) [reply src/lib/db/suggest-change-store.ts:49]: Thanks for the clarification!

File: src/lib/db/suggest-change-store.ts
Comment (thomasheartman) [inline src/lib/db/suggest-change-store.ts:93]: Out of curiosity: How do we usually handle cases like this? I'm assuming you're following the same conventions that we use elsewhere in the code base, so it's probably perfectly on brand.

On a personal level, though, I prefer validating values before they're being sent into functions, so that function don't have to check for invalid arguments. So the function would be like this instead:

```suggestion
    mapRow(row: ISuggestChangeSetRow): ISuggestChangeSet {
```

That said, this is more to facilitate a discussion than to point to something we need to change. I also think there's value in keeping with patterns that we use elsewhere in the code, so there's that 💁🏼

File: src/lib/db/suggest-change-store.ts
Comment (thomasheartman) [inline src/lib/db/suggest-change-store.ts:78]: Does knex have a `first` method or something? DB connectors I've used in the past tend to have a function to get the first or unique item for a given query (or ID) instead of a bunch of rows.

File: src/lib/services/suggest-change-service.ts
Comment (thomasheartman) [inline src/lib/services/suggest-change-service.ts:62]: Do we use module.exports in the code base? I can't remember seeing that elsewhere. Is there a reason you both `export default class` and `module.exports`?

File: src/lib/types/events.ts
Comment (thomasheartman) [inline src/lib/types/events.ts:85]: Nice! Maybe I can ask you to add this to the docs too? A short description of when it fires and what it is in `website/docs/api/admin/events-api.md`? Along with a note that it is an experimental feature? (I'll make it so that we can feature gate it in the future!)

**Comment by thomasheartman** [comment]
Nice work! I've got some more questions if you don't mind 😇

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (FredrikOseberg) [inline src/migrations/20221810114644-add-suggest-changes-table.js:11]: Should this be a JSON B column?

File: src/lib/services/suggest-change-service.ts
Comment (FredrikOseberg) [inline src/lib/services/suggest-change-service.ts:31]: Do we need a method to record events after it has been created? Does creating a changeRequest always require a changeset?

**Review by FredrikOseberg — CHANGES_REQUESTED** [review]
Initial draft makes sense. I've added some comments / questions for clarification purposes.

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (FredrikOseberg) [reply src/migrations/20221810114644-add-suggest-changes-table.js:11]: Or should change events be modeled as it's own table?

File: src/lib/services/suggest-change-service.ts
Comment (thomasheartman) [reply src/lib/services/suggest-change-service.ts:59]: I don't think so: Nested promises collapse, so that should be functionally the same, right?

File: src/lib/db/index.ts
Comment (andreas-unleash) [reply src/lib/db/index.ts:35]: sure, ok

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (andreas-unleash) [reply src/migrations/20221810114644-add-suggest-changes-table.js:11]: Could definitely be it's own table, I think that makes more sense

File: src/lib/services/suggest-change-service.ts
Comment (andreas-unleash) [reply src/lib/services/suggest-change-service.ts:31]: A change request requires a changeSet, yes

File: src/lib/services/suggest-change-service.ts
Comment (FredrikOseberg) [inline src/lib/services/suggest-change-service.ts:1]: The service should move to enterprise

Comment (andreas-unleash) [reply src/lib/services/suggest-change-service.ts:1]: The reason is here is to keep the convention (see SegmentService)

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (sjaanus) [inline src/migrations/20221810114644-add-suggest-changes-table.js:11]: Should we use a foreign key for `user` instead if we want to strongly link the change sets to actual user. In future, we are not displaying the creator just as string, but we will need to run queries on it. Having a `created_by: user.username || user.email` will make it error prone and hard to query.

**Review by sjaanus — CHANGES_REQUESTED** [review]

File: src/lib/db/suggest-change-store.ts
Comment (Tymek) [inline src/lib/db/suggest-change-store.ts:10]: `Changeset` should be cammelCased like `ISuggestChangeset`

File: src/lib/types/model.ts
Comment (Tymek) [inline src/lib/types/model.ts:392]: Is it linked with `User`? How do we store info about who created it, because UI has user name and avatar. Maybe some metadata on this can solve it.

File: src/lib/db/suggest-change-store.ts
Comment (andreas-unleash) [reply src/lib/db/suggest-change-store.ts]: Wouldn't first only get 1 row ?- There might be many rows based on the connections, that will need to be aggregated

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (andreas-unleash) [reply src/migrations/20221810114644-add-suggest-changes-table.js:11]: makes sense

File: src/lib/services/suggest-change-service.ts
Comment (FredrikOseberg) [reply src/lib/services/suggest-change-service.ts]: I suspect this is an oversight, and represents technical debt that we should move to enterprise also. We do have a setup for services in enterprise. 

@ivarconr

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (andreas-unleash) [reply src/migrations/20221810114644-add-suggest-changes-table.js:11]: I went through the migrations, almost everywhere `created_by` is set to a text field

Comment (FredrikOseberg) [reply src/migrations/20221810114644-add-suggest-changes-table.js:11]: I guess it depends on the intent of the table, there may not be a lot of other tables where we need to query for drafts per user, where createdat essentially is metadata and the id is the foreign key.

File: src/lib/db/suggest-change-store.ts
Comment (sjaanus) [reply src/lib/db/suggest-change-store.ts]: You are correct, checked the database query and seems reasonable.

File: src/lib/db/suggest-change-store.ts
Comment (FredrikOseberg) [inline src/lib/db/suggest-change-store.ts:227]: I think we should return the entire changeSet object from the create, this will allow the frontend to save a call and use the returning value instead.

File: src/lib/types/model.ts
Comment (FredrikOseberg) [inline src/lib/types/model.ts:378]: intentional?

**Review by FredrikOseberg — CHANGES_REQUESTED** [review]
Nice work, this is very close. Minor comment.

File: website/docs/api/admin/events-api.md
Comment (thomasheartman) [inline website/docs/api/admin/events-api.md:1580]: Nice! Thank you! Is this what the actual event will look like? Does it contain any data or anything? I'd expect there to be a link to the suggestion that was created, at least? This example should show more or less exactly what the users would see when creating a change suggestion.

File: src/lib/types/model.ts
Comment (thomasheartman) [inline src/lib/types/model.ts:376]: Can you have a change set without changes? Or a change set with an empty list of changes for that matter? If not, maybe it makes sense to make it a required, non-empty list? That is, something like `changes: [ISuggestChange, ...ISuggestChange[]]`?

File: src/lib/types/model.ts
Comment (thomasheartman) [inline src/lib/types/model.ts:375]: Why are these two properties optional? Are there any situations where we'll have a change set without knowing who made it or when?

File: src/lib/types/model.ts
Comment (thomasheartman) [inline src/lib/types/model.ts:386]: Does it make sense to have a change without a payload? Maybe there's cases where it's not necessary (like enabling/disabling a toggle)? And as with the change set: why are createdBy and createdAt optional?

File: src/lib/types/model.ts
Comment (thomasheartman) [inline src/lib/types/model.ts:417]: This is another instance where these fields are listed as optional. Nothing new to add; just wanted to create a comment as a marker 🤓

**Review by thomasheartman — COMMENTED** [review]
Looks good! Thanks a lot for the docs 🙌🏼  Got a couple questions regarding optional properties, that I've marked inline and some further clarification regarding the event doc, if you don't mind 😄

File: src/lib/db/suggest-change-store.ts
Comment (thomasheartman) [inline src/lib/db/suggest-change-store.ts:255]: Why do we need to make this conditional? Can we not just return the change set regardless and ignore it, if not?

Comment (thomasheartman) [reply src/lib/db/suggest-change-store.ts:255]: If it's for performance reasons, is there any way to encode that it will always be returned when you pass `shouldReturn`? Seems like a hassle to have to check whether a function returned `void` or an actual object 💁🏼

File: website/docs/api/admin/events-api.md
Comment (andreas-unleash) [reply website/docs/api/admin/events-api.md:1580]: Would you add the id of the change request or the whole request?

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (sjaanus) [inline src/migrations/20221810114644-add-suggest-changes-table.js:10]: Should project and environment be also foreign keys? What will happen to suggest changes when project is deleted?

File: src/test/fixtures/fake-suggest-change-store.ts
Comment (sjaanus) [inline src/test/fixtures/fake-suggest-change-store.ts:37]: Am I missing something or this should be `===` and also for other methods  ?

File: src/lib/types/model.ts
Comment (andreas-unleash) [reply src/lib/types/model.ts:417]: They were made optional because they are added when commiting to the db - I can create more objects (SuggestChangeCreate and SuggestChangesetCreate and SuggestChangeEventCreate) that do not have them, if you think that this is a problem.

File: website/docs/api/admin/events-api.md
Comment (thomasheartman) [reply website/docs/api/admin/events-api.md:1580]: Whatever we want to put in the event. How have you designed the event? Having both would provide the maximum flexibility, but I don't know how large these events would be. At the very least, you should be able to go and see the resource that was created, right?

File: src/lib/types/model.ts
Comment (andreas-unleash) [reply src/lib/types/model.ts:375]: see below

File: src/lib/types/model.ts
Comment (andreas-unleash) [reply src/lib/types/model.ts:386]: see below

File: src/lib/types/model.ts
Comment (thomasheartman) [reply src/lib/types/model.ts:417]: Got it. But when it comes out of the DB, it'll always have this data, right? In that case, my vote goes to using more types that are more specific (using TS's fancy type system). That way, you'll always know you have  createdBy/createdAt data when it's coming out of the DB 💁🏼 But I'm willing to be challenged on this.

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (andreas-unleash) [reply src/migrations/20221810114644-add-suggest-changes-table.js:10]: Adding it makes sense however, we are already at 5 joins for the query, I am reluctant to add more

File: src/lib/db/suggest-change-store.ts
Comment (thomasheartman) [reply src/lib/db/suggest-change-store.ts:255]: I'm not sure I understand: you need the change set returned in both cases you mentioned over? Why would you then return `void`? Or maybe I misunderstood what you meant.

File: src/test/fixtures/fake-suggest-change-store.ts
Comment (andreas-unleash) [reply src/test/fixtures/fake-suggest-change-store.ts]: Good catch, thanks

File: src/lib/db/suggest-change-store.ts
Comment (andreas-unleash) [reply src/lib/db/suggest-change-store.ts:255]: In the `create` method this code adds the individual changes. I don't want to call get after each add - I want to call it when the whole `create` method is done 
```
suggestChangeSet.changes.forEach((change) => {
            this.addChangeToSet(change, id, user, false);
        });
```

Comment (thomasheartman) [reply src/lib/db/suggest-change-store.ts]: Ah, I think I see now; thanks for explaining 😄  Yeah, this doesn't sound like anything this function should care about to me. You already need the change set ID to add the change regardless, so if we think it would cause issues to always return the updated change set, then I'd personally vote to leave it up to the caller after updating.

Comment (FredrikOseberg) [reply src/lib/db/suggest-change-store.ts]: Agree with Thomas on this one

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (FredrikOseberg) [reply src/migrations/20221810114644-add-suggest-changes-table.js:10]: We can always performance test this and make changes later? Having to manually clean up if an environment / project gets deleted sounds like a nightmare to me.

File: src/lib/db/suggest-change-store.ts
Comment (andreas-unleash) [reply src/lib/db/suggest-change-store.ts]: Removed - moved the return to the service layer

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (sjaanus) [reply src/migrations/20221810114644-add-suggest-changes-table.js:10]: Yea, I don't think the performance hit will be something remarkable. Those joins on foreign keys are quite fast.

File: src/test/fixtures/fake-suggest-change-store.ts
Comment (sjaanus) [inline src/test/fixtures/fake-suggest-change-store.ts:29]: This needs `===`

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (chriswk) [reply src/migrations/20221810114644-add-suggest-changes-table.js:10]: Provided you add index on elements that you do FK joins on

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (chriswk) [inline src/migrations/20221810114644-add-suggest-changes-table.js:19]: We really should be using jsonb.

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (chriswk) [inline src/migrations/20221810114644-add-suggest-changes-table.js:28]: Same thing, please use jsonb unless you have really good reason for not doing so.

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (sjaanus) [reply src/migrations/20221810114644-add-suggest-changes-table.js:10]: Primary keys are indexed by default.

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (sjaanus) [reply src/migrations/20221810114644-add-suggest-changes-table.js:19]: Good idea

File: src/migrations/20221810114644-add-suggest-changes-table.js
Comment (sjaanus) [reply src/migrations/20221810114644-add-suggest-changes-table.js:28]: Good idea

File: src/lib/db/suggest-change-store.ts
Comment (sjaanus) [inline src/lib/db/suggest-change-store.ts:155]: What is this secret? I dont see it in migrations.

**Review by sjaanus — APPROVED** [review]
LGTM!

**Review by FredrikOseberg — CHANGES_REQUESTED** [review]
Looks good, though we should get rid of the console.logs before merging.

**Pull request merged**
